ts/store/helpers/indexer.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * IndexedObject helps creating objects that
+ * are indexed by their id (storing both
+ * byId and allIds)
+ */
+
+import { none, Option, some } from "fp-ts/lib/Option";
+
+type Id = string | number;
+
+// an object with an identifier
+// (assuming strings and numbers
+// only as identifiers -- change as needed)
+interface WithId {
+  id: Id;
+}
+
+/**
+ * represents a "list" of objects that have
+ * been indexed by their "id" property
+ */
+interface IndexedById<V extends WithId> {
+  [key: string]: V;
+  [key: number]: V;
+}
+
+/**
+ * An indexed object generated from a list of objects:
+ * Instantiated by IndexedObject.create()
+ *  - input: [ { id: 1, payload: "X" }, { id: 42, payload: "Y" } ]
+ *  - output:
+ *    {
+ *      byId: { 1: { id: 1, payload: "X" }, 42: { id: 42, payload: "Y" } },
+ *      allIds: [1, 42]
+ *    }
+ *  - V: object with an index (named id) (e.g. CreditCard)
+ */
+export class IndexedObject<V extends WithId> {
```

Comment:
this class looks superfluous, get/values/keys can be implemented for any indexed type

also the `create` constructor could be just a method, e.g. `toIndexex<T, V extends<WIthId<T>>>(a: ReadonlyArray<V>): IndexedById<T, V> {...}`

Reply:
How would you recommend extracting a single value, values and keys? I'm familiar with `Object`, but that's literally the first *Don't* in [Typescript's Do's and Don'ts list](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html) so I was wondering whether different type safe alternatives exist

Reply:
I mean, you probably just need this:

```ts
function toIndexed<T, V extends<WIthId<T>>>(a: ReadonlyArray<V>): IndexedById<T, V> {
   return lst.reduce(
     (o, obj) => ({ ...o, [obj.id]: obj }),
     {} as IndexedById<T, V>
   );
 }
```

then get/values/keys can be functions that accept an `IndexedById<T, V>` parameter

Reply:
So just so we're clear, you'd recommend having functions like:
```typescript
function indexedObjectGet(obj: IndexedById<T,V>): Option<V> { ... }
function indexedObjectValues(obj: IndexedById<T,V>): ReadonlyArray<V> { ... }
function indexedObjectKeys(obj: IndexedById<T,V>): ReadonlyArray<T> { ... }
```
correct?

Reply:
yes but I would say you don't need those functions: for getting keys and values you can use Lodash, for optional get you can use `Option.fromNullable(obj[key])`

Reply:
I didn't notice Loadash and didn't want to add other dependencies just for that, but since it's already there it's great news :D